### ISSUE TITLE
Variable not allowed in type expression

### ISSUE THREAD

**Issue by GF-Huang** [microsoft/pyright#9051]
## Environment data

-   Pylance version: `v2024.9.2`
-   OS and version: `Win11 23H2 22631.4169`
-   Python version (& distribution if applicable, e.g. Anaconda): `Conda Python 3.11.9`

## Code Snippet

![image](https://github.com/user-attachments/assets/416e441c-c61d-47bc-8ed9-c4e0961a87d9)

```python
import cupy as cp
from cupy.typing import NDArray as CPArray

x: CPArray = cp.array([])
x
```

## Repro Steps

1. XXX

## Expected behavior

No error reported.

## Actual behavior

Error reported.

## Logs

```
XXX
```

**Comment by erictraut** [comment]
I'm not able to repro this because `cupy` won't install on MacOS. 

@debonte, are you able to repro this on your Windows system? Does it still repro with the latest version of pyright? This smells like a bug that I fixed in 1.1.381.

**Comment by debonte** [comment]
Yes, this repros on Pyright 1.1.381.

**Comment by erictraut** [comment]
This behavior is expected given the way that `cupy` defines `NDArray`. It appears that the library intends to define it as a type alias, but it's using the old-style (PEP 484) way to define the type alias, like this:

```python
NDArray = core.ndarray[Any, numpy.dtype[_ScalarType_co]]
```

This technique for defining a type alias is problematic because it uses the same syntactic form as a simple variable assignment. Type checkers need to use a variety of heuristics to distinguish between a variable assignment and a type alias definition. In this case, pyright isn't able to evaluate the type on the RHS because one or more of the symbols are defined in a native library, and no type stub is provided, so they evaluate to `Unknown`. This prevents pyright from determining whether the `NDArray` symbol should be a variable or a type alias. Pyright defaults to variables in this case because variable assignments are much more common than type alias definitions. If the heuristic were to default to type aliases in this case, it would result many more false positive errors.

To fix this, the authors of this library can do one of the following:
1. Include type stubs (".pyi" files) in their distribution for native submodules that export symbols. This includes the "cupy/_core/core" submodule.
2. Use the `typing.TypeAlias` directive introduced in [PEP 613](https://peps.python.org/pep-0613) to clearly define this symbol as a type alias. This will prevent pyright (and other type checkers) from assuming this symbol is a variable, but without option 1, the type alias will have little or no utility because it will represent some type that is unknown at static analysis time.

If this library is important to you and you want to use it with type checking, I recommend reporting this issue to the library's maintainers and/or submitting a PR to fix this issue.

As a user of this library, your options are limited if this issue isn't fixed in the library. Here are some suggestions:
1. Import the NDArray symbol from numpy rather than from cupy. I presume they are the same type, and cupy is simply re-exporting it. You could import it from numpy under an `if TYPE_CHECKING` conditional so you get the cupy type at runtime but the numpy type at type checking time.
2. Use a `# pyright: ignore` or `# type: ignore` directive to suppress this error.

I'm going to close this issue because pyright is working as intended here, and there's nothing actionable for us to do.
